src/nlp_expr.jl

```diff
@@ -0,0 +1,156 @@
+        push!(expr.nodes, MOI.Nonlinear.Node(node_type, id, parent))
+        parent = length(expr.nodes)
+        MOI.Nonlinear.parse_expression(data, expr, x.args[1], parent)
+        return
+    end
+    id = get(data.operators.multivariate_operator_to_id, x.head, nothing)
+    @assert id !== nothing
+    node_type = MOI.Nonlinear.NODE_CALL_MULTIVARIATE
+    push!(expr.nodes, MOI.Nonlinear.Node(node_type, id, parent))
+    parent = length(expr.nodes)
+    for arg in x.args
+        MOI.Nonlinear.parse_expression(data, expr, arg, parent)
+    end
+    return
+end
+
+# Method definitions
+
+Base.zero(::Type{NonlinearExpr}) = NonlinearExpr(:+, [0.0])
+
+Base.one(::Type{NonlinearExpr}) = NonlinearExpr(:+, [1.0])
+
+# Univariate operators
+
+for f in MOI.Nonlinear.DEFAULT_UNIVARIATE_OPERATORS
+    op = Meta.quot(f)
+    if f in (:+, :-)
+        continue  # We don't need these.
+    elseif isdefined(Base, f)
+        @eval Base.$(f)(x::AbstractJuMPScalar) = NonlinearExpr($op, [x])
+    else
+        # The operator is defined in some other package.
+    end
+end
+
+# Multivariate operators
+
+# The multivariate operators in MOI are +, -, *, ^, /, ifelse, atan
+#
+# However, ifelse is a builtin, so we can't add methods to it.
+
+# We need only very generic fallbacks for these, because all other cases are
+# caught with more specific methods.
+for f in (:+, :-, :*, :/, :^, :atan)
+    op = Meta.quot(f)
+    @eval begin
+        function Base.$(f)(x::AbstractJuMPScalar, y::_Scalar)
+            return NonlinearExpr($op, [x, y])
+        end
+        function Base.$(f)(x::_Scalar, y::AbstractJuMPScalar)
+            return NonlinearExpr($op, [x, y])
+        end
+        function Base.$(f)(x::AbstractJuMPScalar, y::AbstractJuMPScalar)
+            return NonlinearExpr($op, [x, y])
+        end
+    end
+end
+
+# JuMP interop
+
+function set_objective_function(model::Model, expr::NonlinearExpr)
+    nlp = nonlinear_model(model; force = true)
+    MOI.Nonlinear.set_objective(nlp, expr)
+    return
+end
+
+function add_constraint(
+    model::Model,
+    con::ScalarConstraint{NonlinearExpr,S},
+    ::String,  # Name is not used in nonlinear constraints.
+) where {S}
+    nlp = nonlinear_model(model; force = true)
+    index = MOI.Nonlinear.add_constraint(nlp, con.func, con.set)
+    return ConstraintRef(model, index, ScalarShape())
+end
+
+# MutableArithmetics.jl
+
+# These converts are used in the {add,sub}mul definition for AbstractJuMPScalar.
+
+function Base.convert(::Type{NonlinearExpr}, x::AbstractVariableRef)
+    return NonlinearExpr(:+, [x])
+end
+
+function Base.convert(::Type{NonlinearExpr}, x::GenericAffExpr)
+    args = Any[x.constant]
+    for (variable, coef) in x.terms
+        push!(args, NonlinearExpr(:*, [coef, variable]))
+    end
+    return NonlinearExpr(:+, args)
+end
+
+function Base.convert(::Type{NonlinearExpr}, x::GenericQuadExpr)
+    args = Any[convert(NonlinearExpr, x.aff)]
```

Comment:
We could also add the quadratic terms to the expression returned by `convert(NonlinearExpr, x.aff)` directly since we know it is a sum.